README
NAME
    CSS::Tiny - Read/Write .css files with as little code as possible

SYNOPSIS
        # In your .css file
        H1 { color: blue }
        H2 { color: red; font-family: Arial }
        .this, .that { color: yellow }
        
    # In your program
        use CSS::Tiny;
        
    # Create a CSS stylesheet
        my $CSS = CSS::Tiny->new();
        
    # Open a CSS stylesheet
        $CSS = CSS::Tiny->read( 'style.css' );
        
    # Reading properties
        my $header_color = $CSS->{H1}->{color};
        my $header2_hashref = $CSS->{H2};
        my $this_color = $CSS->{'.this'}->{color};
        my $that_color = $CSS->{'.that'}->{color};
        
    # Changing styles and properties
        $CSS->{'.newstyle'} = { color => '#FFFFFF' }; # Add a style
        $CSS->{H1}->{color} = 'black';                # Change a property
        delete $CSS->{H2};                            # Delete a style
        
    # Save a CSS stylesheet
        $CSS->write( 'style.css' );
        
    # Get the CSS as a <style>...</style> tag
        $CSS->html;

DESCRIPTION
    "CSS::Tiny" is a perl class to read and write .css stylesheets with as
    little code as possible, reducing load time and memory overhead. CSS.pm
    requires about 2.6 meg or ram to load, which is a large amount of
    overhead if you only want to do trivial things. Memory usage is normally
    scoffed at in Perl, but in my opinion should be at least kept in mind.

    This module is primarily for reading and writing simple files, and
    anything we write shouldn't need to have documentation/comments. If you
    need something with more power, move up to CSS.pm. With the increasing
    complexity of CSS, this is becoming more common, but many situations can
    still live with simple CSS files.

  CSS Feature Support
    "CSS::Tiny" supports grouped styles of the form "this, that { color:
    blue }" correctly when reading, ungrouping them into the hash structure.
    However, it will not restore the grouping should you write the file back
    out. In this case, an entry in the original file of the form

        H1, H2 { color: blue }

    would become

        H1 { color: blue }
        H2 { color: blue }

    "CSS::Tiny" handles nested styles of the form "P EM { color: red }" in
    reads and writes correctly, making the property available in the form

        $CSS->{'P EM'}->{color}

    "CSS::Tiny" ignores comments of the form "/* comment */" on read
    correctly, however these comments will not be written back out to the
    file.

CSS FILE SYNTAX
    Files are written in a relatively human-orientated form, as follows:

        H1 {
            color: blue;
        }
        .this {
            color: red;
            font-size: 10px;
        }
        P EM {
            color: yellow;
        }

    When reading and writing, all property descriptors, for example "color"
    and "font-size" in the example above, are converted to lower case. As an
    example, take the following CSS.

        P {
            Font-Family: Verdana;
        }

    To get the value 'Verdana' from the object $CSS, you should reference
    the key "$CSS->{P}->{font-family}".

METHODS
  new
    The constructor "new" creates and returns an empty "CSS::Tiny" object.

  read $filename
    The "read" constructor reads a CSS stylesheet, and returns a new
    "CSS::Tiny" object containing the properties in the file.

    Returns the object on success, or "undef" on error.

  read_string $string
    The "read_string" constructor reads a CSS stylesheet from a string.

    Returns the object on success, or "undef" on error.

  clone
    The "clone" method creates an identical copy of an existing "CSS::Tiny"
    object.

  write_string
    Generates the stylesheet for the object and returns it as a string.

  write
    The "write $filename" generates the stylesheet for the properties, and
    writes it to disk. Returns true on success. Returns "undef" on error.

  html
    The "html" method generates the CSS, but wrapped in a "style" HTML tag,
    so that it can be dropped directly onto a HTML page.

  xhtml
    The "html" method generates the CSS, but wrapped in a "style" XHTML tag,
    so that it can be dropped directly onto an XHTML page.

  errstr
    When an error occurs, you can retrieve the error message either from the
    $CSS::Tiny::errstr variable, or using the "errstr" method.

CAVEATS
  CSS Rule Order
    While the order of rules in CSS is important, this is one of the
    features that is sacrificed to keep things small and dependency-free. If
    you need to preserve order yourself, we recommend that you upgrade to
    the more powerful CSS module.

    If this is not possible in your case, alternatively it can be done with
    the help of another module such as Tie::IxHash:

        my $css = CSS::Tiny->new;
        tie %$css, 'Tie::IxHash';
        $css->read('style.css');

    Note: You will also need to remember to add the additional dependency to
    your code or module in this case.

SUPPORT
    Bugs should be reported via the CPAN bug tracker at

    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=CSS-Tiny>

    For other issues, or commercial enhancement or support, contact the
    author.

AUTHOR
    Adam Kennedy <adamk@cpan.org>

SEE ALSO
    CSS, <http://www.w3.org/TR/REC-CSS1>, Config::Tiny, <http://ali.as/>

COPYRIGHT
    Copyright 2002 - 2010 Adam Kennedy.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Perl extension CSS-Tiny

1.20 Thu 16 Jan 2014
	- Modified FSF address (#73911)
	- Now deals with properties starting with asterisk (#87261)
	- Resolved xt/ tests failing due to incorrect declaration of
	  compatible perl version. Now requires 5.005  (#90611)

1.19 Fri  3 Sep 2010
	- No functional changes
	- Resolved RT #59037: CSS::Tiny reorders rules
	- The use of Tie::IxHash to achieve ordering is now documented

1.18 Fri  3 Sep 2010
	- Resolved RT #60776: Parsing of multiple whitespace selectors
	- Added test script specifically for RT regression tests

1.17 Sun  4 Jul 2010
	- Removed the use of Module::Install
	- Allow read-string to be rerun again on an existing object,
	  and thus also thus on tied objects (MPETERS)
	- Moved POD after __END__ to reduce number of disk reads when parsing.
	- Added support for running xt tests

1.16 Fri  4 Jul 2008
	- Upgrading to Module::Install 0.76
	- Localising $@ during evals

1.15 Thu  1 Nov 2007
	- Moving from Module::Install to the tinier EU:MM
	- Changed use 5.005 to BEGIN { require 5.004 }

1.14 Mon  4 Sep 2006
	- Upgrading to Module::Install 0.64

1.12 Thu  9 Jun 2006
	- Moved from old CVS to new SVN repository
	- Removed uses of UNIVERSAL::isa
	- Upgraded to Module::Install 0.63

1.11 Sat Oct 22 2005
	- Fixed a bug in the inline ->clone method that only runs
	  when Clone is unavailable.
	- Dialed back the File::Spec dep a little to prevent 5.6 people
	  needing to upgraded when not needed.

1.10 Thu May 26 2005
	- Added the ->clone method

1.09 Thu Mar 24 2005
	- Lars Thegler noted that 3-argument open is not supported by 5.005.
	  Added a small fix to change it to 2-argument open.

1.08 Fri Feb 18 2005
	- Adding XHTML support with the ->xhtml method
	- Cleaning up the POD a little

1.07 Tue Feb 15 2005
	- Removing braindead Build.PL

1.06 Sun Jan 16 2005
	- Added ->html to generate CSS to dump on a HTML page
	- Modernised the test suite and makefile

1.05 Wed Jul 14 2004
	- Switched from open( FILE, ">filename" ) to open( FILE, '>', 'filename' )
	- Using the -X _ method to file-test via cached stat information
	- Added support for merging differently named properties in multiply listed styles
	- Upgraded test script to hopefully work on Win32 now

1.04 Wed Jan  7 2004
	- Changed property parser to no longer require exactly one space
	  after the colon seperate. Now allows zero or more spaces.

1.02 Wed Aug 13 16:12:51 2003
	- Applied a series of small changes to reduce the code overhead
	  by a K or so, without changing function

1.01 Mon Jul 15 00:28:12 2003
	- Changed write_string to sort in reverse order. This fixes
	  a problem with A:hover described at the url
	  http://www.w3.org/TR/CSS2/selector.html#dynamic-pseudo-classes
	  Fix supplied by Blaise Barrelet

1.0  Sat Dec 21 12:31:51 2002
	- Removed locking code, and reliance on Fcntl
	- Removed support for write modes
	- Tightened up the code
	- Added the read_string() method
	- Simplified the file generator to save ops

0.3  Mon Dec 09 00:27:01 2002
	- Upgraded tests to use Test::More and compare structs

0.2  Thu Nov 28 16:53:21 2002
	- Regex to remove C style comments was broken
	  Fix supplied by Bjorn Svensson.

0.1  Tue Nov 26 20:53:23 2002
	- original version


Makefile.PL
use strict;
use vars qw{$VERSION};
BEGIN {
	require 5.005;
	$VERSION = '1.20';
}
use ExtUtils::MakeMaker;

WriteMakefile(
	MIN_PERL_VERSION => '5.005',
	META_MERGE => {
		resources => {
			repository => 'http://svn.ali.as/cpan/trunk/CSS-Tiny',
		},
	},
	NAME      => 'CSS::Tiny',
	ABSTRACT  => 'Read/Write .css files with as little code as possible',
	VERSION   => $VERSION,
	PREREQ_PM => {
	},
	BUILD_REQUIRES => {
		# Skip on Windows to avoid breaking ActivePerl PPMs
		# 0.47 means 5.6.2 or newer, which everyone on Win32 has.
		($^O eq 'MSWin32' ? () : ('Test::More' => '0.47')),
	},
	AUTHOR  => 'Adam Kennedy <adamk@cpan.org>',
	LICENSE => 'perl',
	(-d 'xt' and $ENV{AUTOMATED_TESTING} || $ENV{RELEASE_TESTING}) ? (
		test => {
			TESTS => 't/*.t xt/*.t',
		},
	) : (),
);

sub WriteMakefile {  #Written by Alexandr Ciornii, version 0.21. Added by eumm-upgrade.
	my %params=@_;
	my $eumm_version=$ExtUtils::MakeMaker::VERSION;
	$eumm_version=eval $eumm_version;
	if ( $params{BUILD_REQUIRES} and $eumm_version < 6.5503 ) {
		#EUMM 6.5502 has problems with BUILD_REQUIRES
		$params{PREREQ_PM} = {
			%{$params{PREREQ_PM} || {}},
			%{$params{BUILD_REQUIRES}}
		};
		delete $params{BUILD_REQUIRES};
	}
	delete $params{MIN_PERL_VERSION} if $eumm_version < 6.48;
	delete $params{META_MERGE} if $eumm_version < 6.46;
	delete $params{LICENSE} if $eumm_version < 6.31;
	delete $params{AUTHOR} if $] < 5.005;

	ExtUtils::MakeMaker::WriteMakefile(%params);
}


lib/CSS/Tiny.pm
package CSS::Tiny;

# See POD at end for docs

use strict;
BEGIN {
	require 5.004;
	$CSS::Tiny::VERSION = '1.19';
	$CSS::Tiny::errstr  = '';
}

# Create an empty object
sub new { bless {}, shift }

# Create an object from a file
sub read {
	my $class = shift;

	# Check the file
	my $file = shift or return $class->_error( 'You did not specify a file name' );
	return $class->_error( "The file '$file' does not exist" )          unless -e $file;
	return $class->_error( "'$file' is a directory, not a file" )       unless -f _;
	return $class->_error( "Insufficient permissions to read '$file'" ) unless -r _;

	# Read the file
	local $/ = undef;
	open( CSS, $file ) or return $class->_error( "Failed to open file '$file': $!" );
	my $contents = <CSS>;
	close( CSS );

	$class->read_string( $contents )
}

# Create an object from a string
sub read_string {
	my $self = ref $_[0] ? shift : bless {}, shift;

	# Flatten whitespace and remove /* comment */ style comments
	my $string = shift;
	$string =~ tr/\n\t/  /;
	$string =~ s!/\*.*?\*\/!!g;

	# Split into styles
	foreach ( grep { /\S/ } split /(?<=\})/, $string ) {
		unless ( /^\s*([^{]+?)\s*\{(.*)\}\s*$/ ) {
			return $self->_error( "Invalid or unexpected style data '$_'" );
		}

		# Split in such a way as to support grouped styles
		my $style      = $1;
		my $properties = $2;
		$style =~ s/\s{2,}/ /g;
		my @styles = grep { s/\s+/ /g; 1; } grep { /\S/ } split /\s*,\s*/, $style;
		foreach ( @styles ) { $self->{$_} ||= {} }

		# Split into properties
		foreach ( grep { /\S/ } split /\;/, $properties ) {
			unless ( /^\s*(\*?[\w._-]+)\s*:\s*(.*?)\s*$/ ) {
				return $self->_error( "Invalid or unexpected property '$_' in style '$style'" );
			}
			foreach ( @styles ) { $self->{$_}->{lc $1} = $2 }
		}
	}

	$self
}

# Copy an object, using Clone.pm if available
BEGIN { local $@; eval "use Clone 'clone';"; eval <<'END_PERL' if $@; }
sub clone {
	my $self = shift;
	my $copy = ref($self)->new;
	foreach my $key ( keys %$self ) {
		my $section = $self->{$key};
		$copy->{$key} = {};
		foreach ( keys %$section ) {
			$copy->{$key}->{$_} = $section->{$_};
		}
	}
	$copy;
}
END_PERL

# Save an object to a file
sub write {
	my $self = shift;
	my $file = shift or return $self->_error( 'No file name provided' );

	# Write the file
	open( CSS, '>'. $file ) or return $self->_error( "Failed to open file '$file' for writing: $!" );
	print CSS $self->write_string;
	close( CSS );
}

# Save an object to a string
sub write_string {
	my $self = shift;

	# Iterate over the styles
	# Note: We use 'reverse' in the sort to avoid a special case related
	# to A:hover even though the file ends up backwards and looks funny.
	# See http://www.w3.org/TR/CSS2/selector.html#dynamic-pseudo-classes
	my $contents = '';
	foreach my $style ( reverse sort keys %$self ) {
		$contents .= "$style {\n";
		foreach ( sort keys %{ $self->{$style} } ) {
			$contents .= "\t" . lc($_) . ": $self->{$style}->{$_};\n";
		}
		$contents .= "}\n";
	}

	return $contents;
}

# Generate a HTML fragment for the CSS
sub html {
	my $css = $_[0]->write_string or return '';
	return "<style type=\"text/css\">\n<!--\n${css}-->\n</style>";
}

# Generate an xhtml fragment for the CSS
sub xhtml {
	my $css = $_[0]->write_string or return '';
	return "<style type=\"text/css\">\n/* <![CDATA[ */\n${css}/* ]]> */\n</style>";
}

# Error handling
sub errstr { $CSS::Tiny::errstr }
sub _error { $CSS::Tiny::errstr = $_[1]; undef }

1;

__END__

=pod

=head1 NAME

CSS::Tiny - Read/Write .css files with as little code as possible

=head1 SYNOPSIS

    # In your .css file
    H1 { color: blue }
    H2 { color: red; font-family: Arial }
    .this, .that { color: yellow }
    
    # In your program
    use CSS::Tiny;
    
    # Create a CSS stylesheet
    my $CSS = CSS::Tiny->new();
    
    # Open a CSS stylesheet
    $CSS = CSS::Tiny->read( 'style.css' );
    
    # Reading properties
    my $header_color = $CSS->{H1}->{color};
    my $header2_hashref = $CSS->{H2};
    my $this_color = $CSS->{'.this'}->{color};
    my $that_color = $CSS->{'.that'}->{color};
    
    # Changing styles and properties
    $CSS->{'.newstyle'} = { color => '#FFFFFF' }; # Add a style
    $CSS->{H1}->{color} = 'black';                # Change a property
    delete $CSS->{H2};                            # Delete a style
    
    # Save a CSS stylesheet
    $CSS->write( 'style.css' );
    
    # Get the CSS as a <style>...</style> tag
    $CSS->html;

=head1 DESCRIPTION

C<CSS::Tiny> is a perl class to read and write .css stylesheets with as 
little code as possible, reducing load time and memory overhead. CSS.pm
requires about 2.6 meg or ram to load, which is a large amount of 
overhead if you only want to do trivial things.
Memory usage is normally scoffed at in Perl, but in my opinion should be
at least kept in mind.

This module is primarily for reading and writing simple files, and anything
we write shouldn't need to have documentation/comments. If you need
something with more power, move up to CSS.pm. With the increasing complexity
of CSS, this is becoming more common, but many situations can still live
with simple CSS files.

=head2 CSS Feature Support

C<CSS::Tiny> supports grouped styles of the form
C<this, that { color: blue }> correctly when reading, ungrouping them into
the hash structure. However, it will not restore the grouping should you
write the file back out. In this case, an entry in the original file of
the form

    H1, H2 { color: blue }

would become

    H1 { color: blue }
    H2 { color: blue }

C<CSS::Tiny> handles nested styles of the form C<P EM { color: red }>
in reads and writes correctly, making the property available in the
form

    $CSS->{'P EM'}->{color}

C<CSS::Tiny> ignores comments of the form C</* comment */> on read
correctly, however these comments will not be written back out to the
file.

=head1 CSS FILE SYNTAX

Files are written in a relatively human-orientated form, as follows:

    H1 {
        color: blue;
    }
    .this {
    	color: red;
    	font-size: 10px;
    }
    P EM {
    	color: yellow;
    }

When reading and writing, all property descriptors, for example C<color>
and C<font-size> in the example above, are converted to lower case. As an
example, take the following CSS.

    P {
    	Font-Family: Verdana;
    }

To get the value C<'Verdana'> from the object C<$CSS>, you should
reference the key C<$CSS-E<gt>{P}-E<gt>{font-family}>.

=head1 METHODS

=head2 new

The constructor C<new> creates and returns an empty C<CSS::Tiny> object.

=head2 read $filename

The C<read> constructor reads a CSS stylesheet, and returns a new
C<CSS::Tiny> object containing the properties in the file.

Returns the object on success, or C<undef> on error.

=head2 read_string $string

The C<read_string> constructor reads a CSS stylesheet from a string.

Returns the object on success, or C<undef> on error.

=head2 clone

The C<clone> method creates an identical copy of an existing C<CSS::Tiny>
object.

=head2 write_string

Generates the stylesheet for the object and returns it as a string.

=head2 write

The C<write $filename> generates the stylesheet for the properties, and 
writes it to disk. Returns true on success. Returns C<undef> on error.

=head2 html

The C<html> method generates the CSS, but wrapped in a C<style> HTML tag,
so that it can be dropped directly onto a HTML page.

=head2 xhtml

The C<html> method generates the CSS, but wrapped in a C<style> XHTML tag,
so that it can be dropped directly onto an XHTML page.

=head2 errstr

When an error occurs, you can retrieve the error message either from the
C<$CSS::Tiny::errstr> variable, or using the C<errstr> method.

=head1 CAVEATS

=head2 CSS Rule Order

While the order of rules in CSS is important, this is one of the features
that is sacrificed to keep things small and dependency-free. If you need
to preserve order yourself, we recommend that you upgrade to the more
powerful L<CSS> module.

If this is not possible in your case, alternatively it can be done with the
help of another module such as L<Tie::IxHash>:

    my $css = CSS::Tiny->new;
    tie %$css, 'Tie::IxHash';
    $css->read('style.css');

Note: You will also need to remember to add the additional dependency to
your code or module in this case.

=head1 SUPPORT

Bugs should be reported via the CPAN bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=CSS-Tiny>

For other issues, or commercial enhancement or support, contact the author.

=head1 AUTHOR

Adam Kennedy E<lt>adamk@cpan.orgE<gt>

=head1 SEE ALSO

L<CSS>, L<http://www.w3.org/TR/REC-CSS1>, L<Config::Tiny>, L<http://ali.as/>

=head1 COPYRIGHT

Copyright 2002 - 2010 Adam Kennedy.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut
